cbmc_viewer/clauset.py

```diff
@@ -0,0 +1,264 @@
+                                                      self.instrNotInCore)
+        util.dump(clause_html, filename or "clause.html", outdir)
+
+################################################################
+# Json key tags to access elements in cbmc json output
+
+JSON_GOTO_ID_KEY = 'GOTO_ID'
+JSON_SOURCE_LOC_KEY = 'Source'
+JSON_SAT_HARDNESS_KEY = 'SAT_hardness'
+JSON_CLAUSES_KEY = 'Clauses'
+JSON_CLAUSE_SET_KEY = 'ClauseSet'
+
+# Example cbmc json output
+# [
+#   {
+#     "GOTO": "// Labels: __CPROVER_HIDE__CPROVER_initialize();",
+#     "GOTO_ID": 0,
+#     "SAT_hardness": {
+#       "ClauseSet": [ ],
+#       "Clauses": 0,
+#       "Literals": 0,
+#       "Variables": 0
+#     },
+#     "SSA_expr": "true",
+#     "Source": {}
+#   },
+#   ...
+#   {
+#     "GOTO": "ASSERT main::1::a[cast(1, signedbv[64])] < 10",
+#     "GOTO_ID": 95,
+#     "SAT_hardness": {
+#       "ClauseSet": [ 6, 7, 8 ... ],
+#       "Clauses": 134,
+#       "Literals": 459,
+#       "Variables": 67
+#     },
+#     "SSA_expr": "¬(main::1::a!0@1#3[1] ≥ 10)",
+#     "Source": {
+#       "file": "simple_main.c",
+#       "function": "main",
+#       "line": "21",
+#       "workingDirectory": "..."
+#     }
+#   }
+# ]
+
+################################################################
+# Utility functions to generate clause summary
+
+def hash(clause):
+    # returns string format of clause to be used as hash key
+    clause_hash_key = ''
+    for lit in clause:
+        clause_hash_key += str(lit) + ' '
+
+    return clause_hash_key
+
+def get_loc_not_in_core(summary, core_instr):
+    instrNotInCore = []
+    coreSourceList = []
+    for instr in summary:
+        if instr[JSON_GOTO_ID_KEY] in core_instr:
+            coreSourceList.append(instr[JSON_SOURCE_LOC_KEY])
+
+    for instr in summary:
+        if instr[JSON_SAT_HARDNESS_KEY][JSON_CLAUSES_KEY] > 0:
+            if instr[JSON_SOURCE_LOC_KEY] not in coreSourceList:
+                if instr[JSON_SOURCE_LOC_KEY] not in instrNotInCore:
+                    instrNotInCore.append(instr[JSON_SOURCE_LOC_KEY])
+
+    return instrNotInCore
+
+def get_instr_in_core(core_file, clause_hash, instr_length):
+    core_clauses = []
+    instr_list = []
+    keys = clause_hash.keys()
+
+    try:
+        with open(core_file) as f:
+            for line in f.readlines():
+                if "p cnf" in line:
+                    continue
+
+                clause = [int(lit) for lit in line.split(' ')]
+                core_clauses.append(clause)
+
+                clause.sort()
+                clause_hash_key = hash(clause)
+
+                if clause_hash_key in keys:
+                    instr = clause_hash[clause_hash_key]
+                    for i in instr:
+                        instr_list.append(i)
+
+    except IOError as err:
+        fail("Core file not found: {}"
+            .format(core_file))
+
+    instr_list = list(set(instr_list))
+    instr_list.sort()
+
+    core_instr = {
```

Comment:
Is this a single core instruction or a set of instructions in the core?

Reply:
It's the set of instructions in core.